### PR TITLE
More work on forward chaining

### PR DESCRIPTION
**Pull request by misgeatgit**

### CODE REVIEW

File: opencog/reasoning/RuleEngine/rule-engine-src/pln/FCMemory.cc [hunk 1]
```diff
@@ -85,9 +85,23 @@ Handle FCMemory::get_cur_target(void) {
...
 			return true;
+		//recursive lookup
+		else if (LinkCast(hi)) {
+			HandleSeqSeq hseqs;
+			hseqs.push_back(as_->getOutgoing(hi));
+			do {
+				HandleSeq iset = hseqs[hseqs.size() - 1];
+				hseqs.pop_back();
+				for (Handle i : iset) {
+					if (i.value() == h.value())
```
Comment (linas) [inline opencog/reasoning/RuleEngine/rule-engine-src/pln/FCMemory.cc]: It should be enough to say `if(i==h)`  -- you don't need to manually get the value.

File: opencog/reasoning/RuleEngine/rule-engine-src/pln/FCMemory.cc [hunk 2]
```diff
@@ -85,9 +85,23 @@ Handle FCMemory::get_cur_target(void) {
...
+						return true;
+					else if (LinkCast(i))
+						hseqs.push_back(as_->getOutgoing(i));
+				}
+			} while (not hseqs.empty());
+		}
 	}
```
Comment (linas) [inline opencog/reasoning/RuleEngine/rule-engine-src/pln/FCMemory.cc]: FYI, there are a variety of  simple minor utilities  in PatternUtils.h that do similar things -- finding atoms in trees of other atoms, etc.

**Comment by linas** [comment]
I am merging these changes, but I am not really reviewing them closely .. perhaps @AmeBel should be doing this, instead of me?

**Comment by bgoertzel** [comment]
Amen understands what the rule engine, pattern matcher and PLN are doing,
but he isn't a C++ programmer... William might be a better candidate to
review such things...


On Wed, Mar 11, 2015 at 11:48 AM, Linas Vepštas <notifications@github.com>
wrote:

> I am merging these changes, but I am not really reviewing them closely ..
> perhaps @AmeBel <https://github.com/AmeBel> should be doing this, instead
> of me?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/pull/1381#issuecomment-78200613>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Pull request merged**